src/lava/lib/optimization/utils/generators/mis.py

```diff
@@ -16,48 +15,94 @@ class MISProblem:
     convert them to a QUBO formulation.
     """
 
-    def __init__(self, num_vertices: int, connection_prob: float,
-                 seed: int = None):
+    def __init__(self, adjacency_matrix: np.ndarray):
         """
-        Instantiate a new random MIS problem, with the given number of
-        vertices and connection probability.
+        Instantiate a new MIS problem, given the adjacency matrix of a graph.
 
         Parameters
         ----------
-        num_vertices: int
-            Number of vertices of the random graph
-        connection_prob: float
-            Connection probability
-        seed: int, optional
-            Seed for numpy random generator
+        adjacency_matrix:
+            Adjacency matrix of a graph.
         """
-        self._num_vertices = num_vertices
-        self._connection_prob = connection_prob
-        self._seed = seed
-        self._adjacency = self._get_random_graph_mtx()
+        if adjacency_matrix.shape[0] != adjacency_matrix.shape[1]:
+            raise ValueError("The adjacency matrix has to be a square matrix.")
+        self._adjacency = adjacency_matrix
+
+    @property
+    def adjacency_matrix(self) -> np.ndarray:
+        """Returns the adjacency matrix of the graph."""
+        return self._adjacency
 
     @property
     def num_vertices(self) -> int:
         """Returns the number of vertices in the graph."""
-        return self._num_vertices
+        return self._adjacency.shape[0]
 
     @property
-    def connection_prob(self) -> float:
-        """Returns the connection probability of the graph."""
-        return self._connection_prob
+    def num_edges(self) -> int:
+        """Returns the number of edges in the graph."""
+        return np.count_nonzero(self._adjacency)
+
+    @classmethod
+    def from_random_uniform(

```

Comment:
One thing that seems a bit weird to me:
As a user, I would expect that I create a new instance of an MIS using something like
`mis = MIS(...)`
Here, I have very dissimilar options.
For a random graph, I need to call MIS.from_random_uniform().
For a graph with a given adjacency matrix, I call MIS(adjacency_matrix).
That feels a bit inconsistent to me.

Just a minor point. But maybe worth to add a flag `create_random` to the __init__ parameters?

Reply:
The design pattern behind this choice is that of factory methods. In general, MIS is described by its adjacency matrix, and this is what it is used in the constructor. However, it is convenient to have utility methods that generate different kinds of graphs and instantiate the associated MIS problem. 

Compared to having all this logic in the constructor, using factory methods makes it easier to add additional generators later on.